### PR TITLE
Add CIRCT Dockerfile

### PR DESCRIPTION
**Pull request by seldridge**
Adds a Dockerfile for an image consisting of an installation of
llvm/circt off the 'main' branch (whatever it pulls when the
Dockerfile runs). This provides a lean-ish (~500MB) install of
circt-opt, circt-translate, and firtool that can either be wrapped in
a shell script to fake an install of these tools or can be used as a
base image.

Adds nightly publishing of this CIRCT image.

@teqdruid: I'm not sure where this should live. E.g., should this be in llvm/circt proper?

This is trying to solve the issue of I'm developing a dependent project that needs to have access to the binary tools that llvm/circt provides. I obviously don't want to build this in CI myself and cache it, so I created this. (I also don't think such a thing already exists?)

### CODE REVIEW

**Comment by teqdruid** [comment]
> I'm not sure where this should live. E.g., should this be in llvm/circt proper?

As usual, there are pros and cons.

llvm/circt: In general, I like putting distribution stuff in the same repo as the code under the assumption that they'll be a small number of distribution types. This can get obscene quickly. Given that docker images are meant to be usable anywhere, there should only be one. A disadvantage is permissions and ownership: I don't know if the llvm org grants publishing permissions to the repos they own.

Here: we control the circt repo. If we run into issues, we can fix it ourselves instead of having to go through the llvm admins. We'd also have more freedom to pull in sources from other repos which may be useful. For example, a validated version of Verilator. Also, ESI requires capnp to be installed at compile time to be useful but it also provides some utilities which make it useful at runtime.

It's up to you, but I'd be happy for it to live here.

File: circt/Dockerfile
Comment (teqdruid) [inline circt/Dockerfile]: Given that the llvm/circt builds are based on Ubuntu 20.04, we should probably stick with that. Actually, why not base this on the integration image? That way we'd get versions of Verilator, capnp, and yosys which have already been "validated" by the circt integration tests.

File: circt/README.md
Comment (teqdruid) [inline circt/README.md:1]: Docker run commands are soooo boilerplate and cumbersome. You may want to provide some scripts like this example. For instance, start the docker container, run a particular command, and ["do the right thing"](https://github.com/llvm/circt/blob/main/utils/run-docker.sh).

File: circt/Dockerfile
Comment (teqdruid) [inline circt/Dockerfile]: Docker stores a temporary image for each `RUN` so it can very quickly replay steps which it's already cached. The problem with that is it cannot tell if there are changes from the outside world which affect steps. So it's *possible* for this Dockerfile to run from an older cached version instead of re-running this script. I don't think this would happen in the GH workflow since I don't think they cache images between runs, but this behavior may change in the future for efficiency's sake. (Pulling down Docker images for each of the builds which happen in docker containers is probably quite costly.)

I would suggest adding a git commit hash parameter to this Dockerfile indicating the circt commit you want to build, inlining the circt.sh script, and cloning the requested commit. In the workflow, you'd have to get the commit from circt/main and pass it into `docker build`

File: circt/circt.sh
Comment (teqdruid) [inline circt/circt.sh]: I'd suggest using [the script](https://github.com/llvm/circt/blob/main/utils/build-llvm.sh) in circt purely for consistency's sake. We've validated when building with that script...

**Comment by seldridge** [comment]
Thanks for the feedback, @teqdruid! I think this is fine if it lives here. I'll get this updated to address your comments.

**Comment by lattner** [comment]
Yeah if this goes in the circt repo, please scoot it into a subdirectory of utils

File: circt/Dockerfile
Comment (seldridge) [reply circt/Dockerfile]: Sure, I think that's fine. I had originally thought that this would be a minimal `docker` wrapper around CIRCT utilities, but it's fine if it's beefier. I was going to use this for other GitHub CI and I think packaging everything together (including Verilator) will work better there anyway.

File: circt/Dockerfile
Comment (seldridge) [reply circt/Dockerfile]: Ah, cool. I didn't know this was how `RUN` worked (would've helped if I read the documentation... 😄). I can switch to the git commit hash approach.

There's probably a further optimization of making both the CIRCT and LLVM submodule hashes separate so that the latter can be cached if it didn't change. This is probably a future enhancement, though.

File: circt/README.md
Comment (seldridge) [reply circt/README.md:1]: Will do. I'll probably add one each for `firtool`, `circt-opt`, and `circt-translate`.

File: circt/circt.sh
Comment (seldridge) [reply circt/circt.sh]: Sure. Will update in a rebase.

File: circt/Dockerfile
Comment (teqdruid) [reply circt/Dockerfile]: yeah, I was considering suggesting that you also create an MLIR image that we could base this on, but that's probably unnecessary given that this is a nightly job.

**Review by teqdruid — APPROVED** [review]

File: .github/workflows/circt.yml
Comment (seldridge) [inline .github/workflows/circt.yml:19]: I may have gotten this wrong. It may just be `apt install -y jq`. However, if this fails the GH action, I can fix it up.

**Comment by seldridge** [comment]
@teqdruid: I don't have commit access to this repo, so I'll leave it to you to land this one.

**Comment by teqdruid** [comment]
@stephenneuendorffer @drom @lattner Could one of you guys make Schuyler a member of the circt org?

**Comment by lattner** [comment]
@seldridge If you don't already have llvm commit access, please read the llvm developer policy and follow teh instructions, thx!

**Comment by teqdruid** [comment]
This is for the "circt" organization, not for the "llvm" one.

**Comment by lattner** [comment]
> This is for the "circt" organization, not for the "llvm" one.

Aha, @drom can you do the honors here plz?  Thx

**Comment by drom** [comment]
done

**Pull request merged**
